### PR TITLE
Refactor Breakpoints to be location symmetric

### PR DESCRIPTION
**Pull request by jasonLaster**
### Summary of Changes

This is a follow up refactor. The goal is to provide some abstractions so that BPs are better represented in the components:

1. Components don't know/nor care if they're original or generated
2. Components only care about showing BPs
3. Toggle BP takes a line and column and does the rest
4. The reducer formats a generic BP with the "right" location

### CODE REVIEW

File: src/actions/breakpoints.js [hunk 1]
```diff
@@ -239,3 +239,76 @@ export function setBreakpointCondition(
...
+    });
+
+    if ((bp && bp.loading) || !selectedLocation || !selectedSource) {
+      return;
+    }
+
+    const { sourceId } = selectedLocation;
+
+    if (bp) {
+      // NOTE: it's possible the breakpoint has slid to a column
```
Comment (codehag) [inline src/actions/breakpoints.js]: you can use the equalize location util!

File: src/actions/breakpoints.js [hunk 2]
```diff
@@ -239,3 +239,76 @@ export function setBreakpointCondition(
...
+        // Pass in a function to get line text because the breakpoint
+        // may slide and it needs to compute the value at the new
+        // line.
+        { getTextForLine: l => getTextForLine(this.editor.codeMirror, l) }
+      );
+    }
+  };
+}
+
+export function toggleBreakpointDisabledStatus(line) {
```
Comment (codehag) [inline src/actions/breakpoints.js]: i like this, might need a simpler name like "toggleDisabledBreakpoint"

File: src/actions/breakpoints.js [hunk 3]
```diff
@@ -239,3 +239,76 @@ export function setBreakpointCondition(
...
+    const { breakpoints, selectedLocation } = this.props;
+    const bp = breakpointAtLocation(breakpoints, selectedLocation, { line });
+
+    if ((bp && bp.loading) || !selectedLocation) {
+      return;
+    }
+
+    const { sourceId } = selectedLocation;
+
+    if (!bp) {
```
Comment (codehag) [inline src/actions/breakpoints.js]: i wonder if we can avoid this somehow... 🤔

File: src/components/Editor/index.js [hunk 4]
```diff
@@ -476,75 +475,6 @@ class Editor extends PureComponent {
     return !!this.cbPanel;
   }
 
-  toggleBreakpoint(line, column = undefined) {
```
Comment (codehag) [inline src/components/Editor/index.js]: 👍 niiice

File: src/components/Editor/index.js [hunk 5]
```diff
@@ -661,44 +591,6 @@ class Editor extends PureComponent {
     });
   }
 
-  renderBreakpoints() {
```
Comment (codehag) [inline src/components/Editor/index.js]: woooo!

File: src/actions/breakpoints.js [hunk 6]
```diff
@@ -127,7 +133,6 @@ export function disableBreakpoint(location: Location) {
     const action = {
       type: "DISABLE_BREAKPOINT",
       breakpoint: bp,
-      disabled: true,
```
Comment (jasonLaster) [inline src/actions/breakpoints.js]: yup, not being used...

File: src/actions/tests/helpers/breakpoints.js [hunk 7]
```diff
@@ -1,5 +1,16 @@
...
+const sourceFixtures = {
+  foo1: {
+    source: "function foo1() {\n  return 5;\n}",
+    contentType: "text/javascript"
+  },
+  foo2: {
+    source: "function foo2(x, y) {\n  return x + y;\n}",
+    contentType: "text/javascript"
+  }
+};
```
Comment (jasonLaster) [inline src/actions/tests/helpers/breakpoints.js]: perhaps there's a better way to keep this dummy data...

File: src/components/Editor/Breakpoints.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+Breakpoints.displayName = "Breakpoints";
+
+export default connect(
+  state => ({
+    breakpoints: getVisibleBreakpoints(state),
+    selectedSource: getSelectedSource(state)
+  }),
+  dispatch => bindActionCreators(actions, dispatch)
+)(Breakpoints);
```
Comment (jasonLaster) [inline src/components/Editor/Breakpoints.js]: I like the pattern of splitting out the editor children components. The separation makes it easier to follow

@asolove @codehag

File: src/components/Editor/Breakpoint.js [hunk 9]
```diff
@@ -34,39 +35,29 @@ class Breakpoint extends Component {
...
+    if (!selectedSource || breakpoint.loading) {
       return;
     }
 
-    const location = this.getLocation();
-    const line = location.line - 1;
+    const line = breakpoint.location.line - 1;
 
-    editor.setGutterMarker(
+    showSourceText(editor, selectedSource.toJS());
```
Comment (jasonLaster) [inline src/components/Editor/Breakpoint.js]: This is important for making sure that we're updating the correct codemirror document.

File: src/selectors/breakpointAtLocation.js [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    if (!sameLine) {
+      return false;
+    }
+
+    return location.column === getColumn(column, selectedSource);
+  });
+}
+
+/*
+ *
```
Comment (jasonLaster) [inline src/selectors/breakpointAtLocation.js]: I should include a quick description

File: src/selectors/visibleBreakpoints.js [hunk 11]
```diff
@@ -0,0 +1,30 @@
+import { getSelectedSource } from "../reducers/sources";
+import { isGeneratedId } from "devtools-source-map";
+
+function getLocation(breakpoint, isGeneratedSource) {
+  return isGeneratedSource
+    ? breakpoint.generatedLocation || breakpoint.location
+    : breakpoint.location;
+}
+
+export default function getVisibleBreakpoints(state: OuterState) {
```
Comment (jasonLaster) [inline src/selectors/visibleBreakpoints.js]: There should be a quick comment to explain how this is used

File: src/test/mochitest/browser_dbg-sourcemaps2.js [hunk 12]
```diff
@@ -24,7 +24,7 @@ add_task(function*() {
   yield addBreakpoint(dbg, mainSrc, 4);
   is(getBreakpoints(getState()).size, 1, "One breakpoint exists");
   ok(
-    getBreakpoint(getState(), { sourceId: mainSrc.id, line: 4 }),
+    getBreakpoint(getState(), { sourceId: mainSrc.id, line: 4, column: 2 }),
```
Comment (jasonLaster) [inline src/test/mochitest/browser_dbg-sourcemaps2.js]: not sure why this changed :/

It is probably alright though

File: src/utils/breakpoint/index.js [hunk 13]
```diff
@@ -14,8 +13,7 @@ export function firstString(...args) {
 
 export function locationMoved(location, newLocation) {
   return (
-    location.line !== newLocation.line ||
-    (location.column != null && location.column !== newLocation.column)
+    location.line !== newLocation.line || location.column !== newLocation.column
```
Comment (jasonLaster) [inline src/utils/breakpoint/index.js]: this is a nice clean up

File: src/utils/breakpoint/index.js [hunk 14]
```diff
@@ -46,17 +44,10 @@ export function equalizeLocationColumn(location, referenceLocation) {
...
 ) {
-  const isGeneratedSource = isGeneratedId(selectedLocation.sourceId);
-
   return breakpoints.find(breakpoint => {
-    const location = isGeneratedSource
-      ? breakpoint.generatedLocation
-      : breakpoint.location;
-
-    const sameLine = location.line === line + 1;
+    const sameLine = breakpoint.location.line === line + 1;
```
Comment (jasonLaster) [inline src/utils/breakpoint/index.js]: i love these cleanups.

File: src/actions/breakpoints.js [hunk 15]
```diff
@@ -239,3 +244,52 @@ export function setBreakpointCondition(
...
+  return ({ dispatch, getState, client, sourceMaps }: ThunkArgs) => {
+    const selectedSource = getSelectedSource(getState());
+    const bp = getBreakpointAtLocation(getState(), { line, column });
+
+    if (bp && bp.loading) {
+      return;
+    }
+
+    if (bp) {
+      // NOTE: it's possible the breakpoint has slid to a column
```
Comment (codehag) [inline src/actions/breakpoints.js]: do we want to slide a breakpoint to a column if the user is expecting a line breakpoint?

File: src/actions/tests/helpers/breakpoints.js (hunk 7 above)
Comment (codehag) [inline src/actions/tests/helpers/breakpoints.js]: could it make sense to start using a factory library?

File: src/components/Editor/Breakpoints.js [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+  }
+
+  render() {
+    const { breakpoints, selectedSource, editor } = this.props;
+
+    if (!selectedSource || !breakpoints || selectedSource.get("isBlackBoxed")) {
+      return null;
+    }
+
+    return dom.div(
```
Comment (codehag) [inline src/components/Editor/Breakpoints.js]: the Breakpoint component returns null; so, do we need to wrap it in a div, or just return null, with the mapping done in the body of the function?

File: src/components/Editor/GutterMenu.js [hunk 17]
```diff
@@ -1,47 +1,45 @@
 import { showMenu } from "devtools-launchpad";
 
 export default function GutterMenu({
-  bp,
+  breakpoint,
```
Comment (codehag) [inline src/components/Editor/GutterMenu.js]: :+1:

File: src/selectors/breakpointAtLocation.js [hunk 18]
```diff
@@ -0,0 +1,60 @@
+import { getSelectedSource } from "../reducers/sources";
+import { getBreakpoints } from "../reducers/breakpoints";
+import { isGeneratedId } from "devtools-source-map";
+
+function isGenerated(selectedSource) {
+  const sourceId = selectedSource.get("id");
+  return isGeneratedId(sourceId);
+}
+
+function getColumn(column, selectedSource) {
```
Comment (codehag) [inline src/selectors/breakpointAtLocation.js]: I like this pattern better than keeping the selectors in the reducer. maybe we could have a folder structure here:
```
selectors/
|- breakpoints/
  |- index.js
  |- atLocation.js
```

File: src/selectors/visibleBreakpoints.js (hunk 11 above)
Comment (codehag) [inline src/selectors/visibleBreakpoints.js]: is this so that we can check if a breakpoint is also represented in a generated file? can we use it in all cases?

File: src/test/mochitest/browser_dbg-sourcemaps2.js (hunk 12 above)
Comment (codehag) [inline src/test/mochitest/browser_dbg-sourcemaps2.js]: this might be a regression - before we were making sure that we didn't turn line breakpoints into column breakpoints...

File: src/utils/breakpoint/index.js (hunk 14 above)
Comment (codehag) [inline src/utils/breakpoint/index.js]: :tada: nice!

File: src/actions/tests/helpers/breakpoints.js (hunk 7 above)
Comment (jasonLaster) [inline src/actions/tests/helpers/breakpoints.js]: yeah - that might be a nice cleanup in a separate PR.

File: src/selectors/visibleBreakpoints.js (hunk 11 above)
Comment (jasonLaster) [inline src/selectors/visibleBreakpoints.js]: This will get the breakpoints for a file at the "correct" location regardless of original / generated
